go-controller/pkg/libovsdbops/model_client.go

```diff
@@ -47,43 +47,77 @@ func extractUUIDsFromModels(models interface{}) []string {
 }
 
 // buildMutationsFromFields builds mutations that use the fields as values.
-func buildMutationsFromFields(fields []interface{}, mutator ovsdb.Mutator) []model.Mutation {
-	mutations := make([]model.Mutation, 0, len(fields))
+func buildMutationsFromFields(fields []interface{}, mutator ovsdb.Mutator) ([]model.Mutation, error) {
+	mutations := []model.Mutation{}
 	for _, field := range fields {
-		v := reflect.ValueOf(field)
-		if v.Kind() != reflect.Ptr {
-			panic(fmt.Sprintf("Expected Ptr but got %s", v.Kind()))
-		}
-		if v.IsNil() || v.Elem().IsNil() {
-			continue
-		}
-
-		if m, ok := field.(*map[string]string); ok {
-			// check if all values on m are zero, if so create slice of keys of m and set that to field
-			allEmpty := true
-			keySlice := []string{}
-			for key, value := range *m {
-				keySlice = append(keySlice, key)
-				if len(value) > 0 {
-					allEmpty = false
-					break
+		switch v := field.(type) {
+		case *map[string]string:
+			if v == nil || len(*v) == 0 {
+				continue
+			}
+			if mutator == ovsdb.MutateOperationDelete {
+				// turn empty map values into a mutation to remove the key for
+				// delete mutations
+				removeKeys := make([]string, 0, len(*v))
+				updateKeys := make(map[string]string, len(*v))
+				for key, value := range *v {
+					if value == "" {
+						removeKeys = append(removeKeys, key)
+					} else {
+						updateKeys[key] = value
+					}
 				}
+				if len(removeKeys) > 0 {
+					mutation := model.Mutation{
+						Field:   field,
+						Mutator: mutator,
+						Value:   removeKeys,
+					}
+					mutations = append(mutations, mutation)
+				}
+				if len(updateKeys) > 0 {
+					mutation := model.Mutation{
+						Field:   field,
+						Mutator: mutator,
+						Value:   updateKeys,
```

Comment:
how `updatekeys` different from using `v.Elem().Interface()` ?

Reply:
On one side, `v` might contain both key/value pairs to be removed and keys to be removed and these are filtered out into `updateKeys` and `removeKeys` and then different mutations. On the other side, `v` is a `*map[string]string` as per the type switch above, not a `reflect.Value`